FWCore/Framework/interface/CallbackExternalWork.h

```diff
@@ -0,0 +1,199 @@
+                  record->activityRegistry()->postESModulePrefetchingSignal_.emit(record->key(),
+                                                                                  Base::callingContext());
+                });
+              } catch (cms::Exception& caughtException) {
+                if (not excptr) {
+                  edm::exceptionContext(caughtException, Base::callingContext());
+                  excptr = std::current_exception();
+                }
+              }
+              if (excptr) {
+                Base::taskList().doneWaiting(excptr);
+                return;
+              }
+
+              Base::queue().push(
+                  *group, [this, holder = std::move(holder), serviceToken, record, eventSetupImpl]() mutable {
+                    Base::callingContext().setState(ESModuleCallingContext::State::kRunning);
+                    std::exception_ptr exceptPtr;
+                    try {
+                      convertException::wrap([this, &holder, &serviceToken, &record, &eventSetupImpl] {
+                        ESModuleCallingContext const& context = Base::callingContext();
+                        auto proxies = Base::getTokenIndices();
+                        if (Base::postMayGetProxies()) {
+                          proxies = &((*Base::postMayGetProxies()).front());
+                        }
+                        TRecord rec;
+                        edm::ESParentContext pc{&context};
+                        rec.setImpl(record, Base::transitionID(), proxies, eventSetupImpl, &pc);
+                        ServiceRegistry::Operate operate(serviceToken.lock());
+                        record->activityRegistry()->preESModuleAcquireSignal_.emit(record->key(), context);
+                        struct EndGuard {
+                          EndGuard(EventSetupRecordImpl const* iRecord, ESModuleCallingContext const& iContext)
+                              : record_{iRecord}, context_{iContext} {}
+                          ~EndGuard() {
+                            record_->activityRegistry()->postESModuleAcquireSignal_.emit(record_->key(), context_);
+                          }
+                          EventSetupRecordImpl const* record_;
+                          ESModuleCallingContext const& context_;
+                        };
+                        EndGuard guard(record, context);
+                        (*acquireFunction_)(rec, holder);
+                      });
+                    } catch (cms::Exception& iException) {
+                      iException.addContext("Running acquire");
+                      exceptPtr = std::current_exception();
+                    }
+                    holder.doneWaiting(exceptPtr);
+                  });
+            });
+      }
+
+      WaitingTask* makeExceptionHandlerTask(WaitingTask* produceTask, oneapi::tbb::task_group* group) {
+        return make_waiting_task([this, produceTask, group](std::exception_ptr const* iException) {
```

Comment:
I'm not a fan of the `produceTask` being held by a 'bare' pointer. I'd rather have `produceTask = WaitingTaskHolder(*group, produceTask)` as that will increment the reference count and guarantee the task will be queued even if the task is deleted without ever being run directly.

Reply:
Or maybe just change `WaitingTask* produceTask` to `WaitingTaskHolder produceTask`.

Reply:
I modified the make functions to return WaitingTaskHolders instead of WaitingTask*. The other way made sense before I refactored. But now this is much better.